### PR TITLE
Modbus update v2

### PR DESCRIPTION
**Pull request by glongo**
This patchset reworks commit 88d7c8d,
because it breaks a UT (ModbusParserTest08) in the previous version.

Last PR: https://github.com/inliniac/suricata/pull/1777

### CODE REVIEW

File: src/app-layer-modbus.c [hunk 1]
```diff
@@ -598,6 +598,11 @@ static void ModbusParseReadResponse(ModbusTransaction   *tx,
             goto error;
     }
 
+    tx->data = SCMalloc(count);
```
Comment (inliniac) [inline src/app-layer-modbus.c]: I see multiple places in the code where tx->data is alloc'd, all unconditionally. I worry about the case of overwriting existing allocations and thus memory leaks.

File: src/output-json-modbus.c [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+    json_object_set_new(modbusjs, "function",
+                        json_string(ModbusGetFunctionName(modbustx->function))
+                       );
+    if (modbustx->function == MODBUS_FUNC_DIAGNOSTIC) {
+        json_object_set_new(modbusjs, "subfunction",
+                            json_string(ModbusGetSubFunctionName(modbustx->subFunction))
+                           );
+    }
+    if (modbustx->type & MODBUS_TYP_READ) {
+        json_t *read = json_object();
```
Comment (inliniac) [inline src/output-json-modbus.c]: read (and write below) are common C function names, please use another name here.

File: src/app-layer-modbus.c [hunk 3]
```diff
@@ -208,6 +209,15 @@ int ModbusGetAlstateProgress(void *modbus_tx, uint8_t direction) {
         ((modbus->transaction_max - tx->tx_num) > request_flood))
         return 1;
 
+    /* Check if we have a hole: next transaction is over but this one not.
```
Comment (inliniac) [inline src/app-layer-modbus.c]: what if you miss 2? Or is that not possible?

**Comment by inliniac** [comment]
Commit " json-modbus: log modbus fields in alerts" is lying. It doesn't add anything to alerts.

File: src/app-layer-detect-proto.c [hunk 4]
```diff
@@ -345,7 +345,9 @@ static AppProto AppLayerProtoDetectPPGetProto(Flow *f,
             SCLogDebug("toserver - Probing parser found for source port %"PRIu16, f->sp);
 
             /* found based on source port, so use sp registration */
-            pe2 = pp_port_sp->sp;
+            pe2 = pp_port_sp->dp;
+            StreamTcpPacketSwitchDir(NULL, p);
```
Comment (inliniac) [inline src/app-layer-detect-proto.c]: this use of StreamTcpPacketSwitchDir doesn't make sense. It flips the direction flags for this one packet, but does nothing for the rest of the flow.

File: src/flow.h [hunk 5]
```diff
@@ -43,67 +43,70 @@ typedef struct AppLayerParserState_ AppLayerParserState;
 /* per flow flags */
 
 /** At least on packet from the source address was seen */
-#define FLOW_TO_SRC_SEEN                  0x00000001
```
Comment (inliniac) [inline src/flow.h]: This flow renumbering should really be in it's own commit.

File: src/app-layer-detect-proto.c (hunk 4 above)
Comment (inliniac) [inline src/app-layer-detect-proto.c]: I see the problem of the applayer recognizing that the flow's direction is reversed, but we'll need a better solution for dealing with that.

File: src/app-layer-modbus.c (hunk 3 above)
Comment (regit) [inline src/app-layer-modbus.c]: Right, we should iterate here.  @glongo can you give a look ?
